### ISSUE TITLE
Cannot use RETURNING with 2-level nested ROW expressions in SQL Server

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13120]
In SQL Server, it is currently not possible to use 2-level nested `ROW` expressions inside of the `RETURNING` clause, as the emulation cannot handle it. E.g. this:

```java
Record2<Record2<Integer, Record1<Integer>>, Integer> r2 =
ctx.insertInto(TTriggers())
   .columns(TTriggers_COUNTER())
   .values(0)
   .returningResult(row(TTriggers_ID_GENERATED(), row(TTriggers_ID())), TTriggers_COUNTER())
   .fetchSingle();
```

Produces this query:

```sql
declare @result table (
  [id_generated] int,
  [alias_1659049] record, -- Data type "record" not found
  [counter] int
);
insert into [test].[dbo].[t_triggers] ([counter])
output 
  inserted.[id_generated],
  inserted.[id] [row.id],
  inserted.[counter]
into @result
values (0);
merge into @result [r]
using (
  select
    [test].[dbo].[t_triggers].[id_generated],
    [test].[dbo].[t_triggers].[id] [alias_1659049.id],
    [test].[dbo].[t_triggers].[counter] [alias_92185295]
  from [test].[dbo].[t_triggers]
) [s]
on [r].[id_generated] = [s].[id_generated]
when matched then update set
  [alias_1659049] = [s].[alias_1659049], -- flatten here
  [counter] = [s].[alias_92185295];
select [id_generated], [alias_1659049], [counter] -- flatten here
from @result [r];
```

The nested record needs to be flattened as well:

- [ ] in the `@result` table declaration
- [ ] in the `MERGE .. WHEN MATCHED` clause
- [ ] in the `SELECT .. FROM @result` query

**Comment by lukaseder** [comment]
The internal `Tools::flattenCollection` (and maybe others) isn't recursive.

**Comment by lukaseder** [comment]
After the fix, the generated SQL now looks correct again:

```sql
declare @result table (
  [id_generated] int,
  [id] int,
  [counter] int
);
insert into [test].[dbo].[t_triggers] ([counter])
output inserted.[id_generated], inserted.[id], inserted.[counter]
into @result
values (0);
merge into @result [r]
using (
  select
    [test].[dbo].[t_triggers].[id_generated],
    [test].[dbo].[t_triggers].[id] [alias_4163486],
    [test].[dbo].[t_triggers].[counter] [alias_92185295]
  from [test].[dbo].[t_triggers]
) [s]
on [r].[id_generated] = [s].[id_generated]
when matched then update set
  [id] = [s].[alias_4163486],
  [counter] = [s].[alias_92185295];
select [id_generated], [id], [counter]
from @result [r];
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.5 (https://github.com/jOOQ/jOOQ/issues/13121), and 3.15.9 (https://github.com/jOOQ/jOOQ/issues/13122)
